plugin-server/src/worker/ingestion/event-pipeline/runner.ts

```diff
@@ -79,115 +41,110 @@ export class EventPipelineRunner {
+                this,
+                processedEvent,
+                personContainer
+            )
 
+            const preparedEvent = await this.runStep(prepareEventStep, this, normalizedEvent)
+            if (preparedEvent != null) {
+                return await this.runStep(createEventStep, this, preparedEvent, newPersonContainer)
+            }
+        }
         this.hub.statsd?.increment('kafka_queue.buffer_event.processed_and_ingested', {
             didPersonExistAtStart: String(!!didPersonExistAtStart),
         })
-        return result
     }
 
-    async runAsyncHandlersEventPipeline(event: PostIngestionEvent): Promise<EventPipelineResult> {
+    async runAsyncHandlersEventPipeline(event: PostIngestionEvent) {
         this.hub.statsd?.increment('kafka_queue.event_pipeline.start', { pipeline: 'asyncHandlers' })
         const personContainer = new LazyPersonContainer(event.teamId, event.distinctId, this.hub)
-        const result = await this.runPipeline('runAsyncHandlersStep', event, personContainer)
+        const result = await this.runStep(runAsyncHandlersStep, this, event, personContainer)
         this.hub.statsd?.increment('kafka_queue.async_handlers.processed')
         return result
     }
 
-    private async runPipeline<Step extends StepType, ArgsType extends StepParameters<EventPipelineStepsType[Step]>>(
-        name: Step,
-        ...args: ArgsType
-    ): Promise<EventPipelineResult> {
-        let currentStepName: StepType = name
-        let currentArgs: any = args
-
-        while (true) {
-            const timer = new Date()
-            try {
-                const stepResult = await this.runStep(currentStepName, ...currentArgs)
-
-                this.hub.statsd?.increment('kafka_queue.event_pipeline.step', { step: currentStepName })
-                this.hub.statsd?.timing('kafka_queue.event_pipeline.step.timing', timer, { step: currentStepName })
-
-                if (stepResult) {
-                    ;[currentStepName, currentArgs] = stepResult
-                } else {
-                    this.hub.statsd?.increment('kafka_queue.event_pipeline.step.last', {
-                        step: currentStepName,
-                        team_id: String(this.originalEvent?.team_id),
-                    })
-                    return {
-                        lastStep: currentStepName,
-                        args: currentArgs.map((arg: any) => this.serialize(arg)),
-                    }
-                }
-            } catch (error) {
-                await this.handleError(error, currentStepName, currentArgs)
-                return {
-                    lastStep: currentStepName,
-                    args: currentArgs.map((arg: any) => this.serialize(arg)),
-                    error: error.message,
-                }
-            }
-        }
-    }
+    protected runStep<Step extends (...args: any[]) => any>(step: Step, ...args: Parameters<Step>): ReturnType<Step> {
```

Comment:
What are your thoughts on decorators? (not valid typescript below)

Currently we have sth like this in the call locations
```
const event = await this.runStep(step_fn_name, this, event, personContainer)
```
But it could be easier to read if we had decorators, e.g. 
```
@errorHandling(sendToDLQ: boolean)  // instead of STEPS_TO_EMIT_TO_DLQ_ON_FAILURE where we need to add functions
@measure    // timeout guard + send stats info
function step_fn(...)

\\ call location
const event = step_fn_name(event, personContainer)
```